### ISSUE TITLE
User auth through OAuth 2.0 with PKCE flow

### ISSUE THREAD

**Issue by AlbinoDrake238** [EECE3093C/team-project-jam-frontend#38]
OAuth 2.0 and PKCE flow are authorization methods used to authorize a client when secure information can't be stored. This is generally the best practice for us, sad I didn't realize it early but Spotify docs are a nightmare. 

PKCE authorizes through the use of a code challenger and a code verifier, which I'll touch on in the issue but you should check out https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce to learn more

Essentially as I understand it, we create a random string comprised of Alphanumeric characters as well as "./-_" that ranges from 48 - 128 characters (the range is up to our digression but more is better). This string is what's known as the code verifier. We then do a SHA-256 hash on the verifier and encode that to a base64url (different from a regular base64 encoding). This is the code challenger. When we make our initial call for authorization we send the code challenger to the OAuth tenant (Spotify API in this case) along with the method we used to hash it. This method should always be SHA-256 denoted as 'S256' but there are some exceptions. the tenant then holds returns to the client an authorization code, this code is encoded based on our challenger but exactly how is determined by the tenant. Then when we go to request our access token from the tenant we send the authorization code along with our code verifier. The tenant ensures the verifier can be used to derive the code to make sure we are who we say we are then returns us the access token which gives us the ability to make actual calls against the API.

tl;dr secure way of making calls to an api when you can't store an authorization or secret tokens that requires a key, a value encoded based of the key, and a code encoded based of the value encoded from the key.

Requirements:
- Code Challenge and Code Verifier are generated correctly.
- Response returned to callback url

**Comment by AlbinoDrake238** [comment]
Had some issues with getting the tenant to accept my verifier with a hashing function that uses the Javascript built in library crypto.createHash() so I found a node module called 'pkce-challenger' that generates it for us. This has worked every time for me.
Documentation for the module can be found here: https://www.npmjs.com/package/pkce-challenge

**Comment by AlbinoDrake238** [comment]
Accidentally did all my work in main (I know I'm evil but it's all okay) but sadly there is no PR to link, this issue however is closed.
